### ISSUE TITLE
Paving the way for greater and more powerful automation

### ISSUE THREAD

**Issue by Lawouach** [chaostoolkit/chaostoolkit#74]
Hey all,

Recently, various members of the community have put forth ideas that put good pressure on our current "run once from A-to-B" approach to execute an experiment.

When the Chaos Toolkit project started, we realised we didn't have all the answers and we would likely fail to find the right model from the get go. So we staged a basic, yet effective, approach that we hoped would help get the discussion started around the Chaos Engineering experiment model.

The result is the current specification as we know it. However, it has, what some may see has limitations due to its simplicity. Here are a few of those that were raised:

1. output from an activity is not propagated further down the experiment, meaning you cannot refer to a value generated during your run ([here](https://github.com/chaostoolkit/chaostoolkit-lib/issues/44) and [here](https://github.com/chaostoolkit/chaostoolkit/issues/70))
2. sometimes, you want to run the steady state either only before or after the method and see what happens. Right now, it always runs before *and* after which is not always suitable. As an example, assume your system is in a degraded state and you want to run an experiment that try out a recovery process you've defined (in other word, use Chaos Engineering not to shake your system down but to put it back up). Your steady state would be something like "are we back on track?", if that was executed before the method, it would fail before we even had a chance to run the method! So there seems to be a value in turning off the steady state run either before or after.
3. [applying an experiment to a pool of data cannot be done at once natively](https://github.com/chaostoolkit-incubator/chaostoolkit-aws/issues/22). Say you have a bunch of environments (or other resources...) you want to apply the experiment at once, you have no other choice but write your own script to call handle the matrix and call the toolkit manually for each combination
4. a similar one to the previous one is being able to run a set of activities in a loop (not the entire experiment)
5. being able to [abort the experiment](https://github.com/chaostoolkit/chaostoolkit/issues/57) run as soon as the method emits such a signal would avoid wasting time having the whole experiment going through for no reasons
6. following the previous one, is there value in a [test](https://github.com/chaostoolkit/chaostoolkit/issues/63) mode when we always imagined the Chaos Engineering flow to have value *because* it always goes to the end (black box/closer to user behavior?)
7. do we need to improve the [synchronization of activities](https://github.com/chaostoolkit/chaostoolkit/issues/62)? We are solely time-based now (wait for 5s after the activity...) but could we imagine event-based? Meaning we would have to listen for them.

Overall, those ideas make a lot of sense and could turn the toolkit into something more capable. But, more capabilities always come at a price which is greater complexity leading to potential increased fragility.

I created this thread not to find solutions to each of these issues (please comment on the according tickets for specifics ;)) but so that we, as the Chaos Toolkit community, conduct a clear discussion about where we want to take the next steps.

I thinks some of those could make a 1.0 if they feel appropriate and not rushed. Others will have to wait for a 2.0.

Please make sure to speak your mind :)

Cheers,
- Sylvain

**Comment by 0rc** [comment]
@Lawouach thanks for great summary
I would also add a point of conditional logic to your list (which somewhat correlates with point 5 in your list):
* Be able to run or not run certain steps based on the outputs of previous steps

To me and the use-cases I see the most important would be points 1, 3, 4

**Comment by 0rc** [comment]
one more point:

8. Support `tolerance` for `probe` inside `method`. Perhaps also support tolerance for `action`.

**Comment by 0rc** [comment]
another idea I have for improvement:
9. Multi-stage experiments. In other words, possibility to have multiple methods and steady state hypothesis in a specific order within one file, or perhaps referencing files from the master file

**Comment by Lawouach** [comment]
Hey @0rc, sorry for the delay.

Thank you so much for you input. This is what this issue is all about! :)

It feels that 8. and 9. could be toed to 6. (adding a test mode).


It is an interesting point, the toolkit was meant to be used as a discovery tool (tell me something about my system that I don't know yet). However, as users move on with their experiments and want to run them continuously, having more control over their execution is expected.

While I feel comfortable with adding a certain level of control, I certainly wonder how far we can go with the declarative approach.

In other words, there is a huge discussion to have here and that's exciting :)

**Comment by MrBW** [comment]
As we discussed:

Is there a possibility to execute the so-called probe more than once or to execute it several times for a certain time? So far I have used the HTTP provider, which only sent 1 request at a time. Is there a possibility to execute this multiple times as described and to aggregate the results?
So far, for example, I have used Apache Benchmarking for HTTP requests.

During my last workshop, more people than just me came up with this idea and asked me about it. In my opinion it would make sense to use more than 1 request for verification.

**Comment by Lawouach** [comment]
Just a note to myself this might be of interest for this discussion too https://github.com/chaostoolkit/chaostoolkit-lib/issues/32

**Comment by Devatoria** [comment]
Hello there. I'm actually giving a shot to the chaos toolkit to test some use cases we have at Datadog and it's sure that some of the listed features in this issue are things we definitely need. What's the actual state on your side? Do you think you want to keep it simple or add those features but with the complexity it'll generate? I'd really like to have a quick overview of where the project should go in your mind.

Thank you for the answer and for the work you provided here :)

**Comment by Lawouach** [comment]
Hey @Devatoria,

Thanks for kind words on the project. :)

We are currently having lots of discussions out of band around the toolkit and lots of interests around its direction so you will have to assume that some data points may be missing in what I am saying next :p 

I think in general, users who have reached out to us because they want to invest on the toolkit have expressed the need for features that fall into three baskets IMO:

1. functionalities that are part of the scope and remit of the toolkit itself (basically, anything that is about the experiment itself)
2. functionalities that are more around the visibility, organization and scheduling of chaos experiments running (think organizations and collaboration)
3. functionalities that are not tied to the toolkit but likely more an expression of a specific need

With this in mind, how  the project evolves depends on which category they fall into.

At this stage, the toolkit should (finally) see its v1 being released to make a stand that effort people have put into using the toolkit (and contributing to it) until now will remain. Stability is essential to us.

Future minor versions will see some of points, shown in my initial comment, likely be included in some fashion. Not all of them may make it however.

I don't have a v2 definition in my mind yet. I think, v2  is likely one or two years away and will come after the community has grown more with a clearer idea of where the project could go to.

Some functionalities that we see asked more and more are around collaboration and visibility. Those concerns will likely be under the Chaos Platform umbrella as it is its scope. That project is under this org and you can see its code already but it's a month away from being fully documented to appreciate its scope.

Finally, where we would like to also take the project is towards becoming a respected citizen of the [Open Chaos Initiative](https://openchaos.io/). From a functionality front, it means we plan on finally making  `chaos discover` fully featured in respect of analysing your system to suggest recommendations of experiments you would want to run. But also allow you, if you wanted to, to share aspects of your experiments and learnings, for everyone to benefit from (with the idea that it's all knowledge we may benefit from at an industry level).

I know this is high level, so please let me know if you have more specific questions and I'll do my best to reply :)

**Comment by gtfisher** [comment]
One of the things I looked for was the ability to run multiple experiments for one call to chaos run, by passing a list of files or possible wildcards. I realise this leads to complexity such as sequential or parallel execution

**Comment by Lawouach** [comment]
Might be of interest too https://github.com/chaostoolkit/chaostoolkit-lib/issues/137

**Comment by Lawouach** [comment]
See #148 and #149 for work undetaken.

**Comment by Lawouach** [comment]
Another useful feature to consider. https://github.com/chaostoolkit/chaostoolkit/issues/162
